_posts/2016-05-05-cargo-pillars.md

```diff
@@ -0,0 +1,489 @@
+---
+layout: post
+title: "Cargo: predictable dependency management"
+author: Yehuda Katz
+description: "Cargo makes dependency management in Rust easy and predictable"
+---
+
+Cargo's goal is to make modern application package management a core value of
+the Rust programming language.
+
+That concern is driven by a desire to answer questions like:
+
+1. If I want to use `libgit2` in my project, how easy is it?
+2. If I build my project on a different machine, for a different architecture,
+   in CI or for release, am I building from the same source code?
+3. If I build my project for testing, will my indirect dependencies be compiled
+   with debug symbols? If I build my project for release, will my indirect
+   dependencies be compiled with maximum optimizations? How can I be sure?
+4. If someone published a new version of one of my dependencies after I commit,
+   will my CI environment use the same source code? My production environment?
+5. If I add a new dependency (or upgrade one), can that break the build? Can it
+   affect unrelated dependencies? Under what conditions?
+
+**All of these questions (and many more like them) have one thing in common:
+predictability**. One solution to this problem, common in the systems space, is
+vendoring dependencies -- forking them directly into an application's repository
+-- and then managing them manually. But this comes at a substantial per-project
+cost, since there's more to manage. It also comes at an ecosystem-wide cost,
+since there's less sharing. Making sure you can answer all of the questions
+above, all of the time, is hard work.
+
+Package managers for higher-level languages have shown that by turning
+dependency management over to a shared tool, you can have predictability, easy
+workflows that operate over the entire dependency graph, and increased sharing
+and robustness across the ecosystem. When we started planning Rust 1.0, we knew
+we wanted to bring these ideas to a systems setting, and making Cargo a central
+part of the way people use Rust was a big part of that.
+
+## Pillars of Cargo
+
+Cargo is built on three major pillars:
+
+1. Building, testing, and running projects should be predictable across
+   environments and over time.
+
+2. To the extent possible, indirect dependencies should be invisible to
+   application authors.
+
+3. Cargo should provide a shared workflow for the Rust ecosystem that aids the
+   first two goals.
+
+We'll look at each of these pillars in turn.
+
+## Predictability
+
+Cargo's predictability goals start with a simple guarantee: **once a
+project successfully compiles, subsequent compiles will use exactly the same
+source code, across machines and environments**.
```

Comment:
nit: wording here is awkward.  I don't want subsequent compiles to use the same project source over and over.  

If I understand correctly, you could use: "once a project successfully compiles, subsequent compiles will use exactly the same source for all dependencies, across machines and environments"